src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs

```diff
@@ -0,0 +1,228 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Logging;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    /// <summary>
+    ///     Applies <see cref="IProjectVersionedValue{T}"/> values to a <see cref="IWorkspaceProjectContext"/>.
+    /// </summary>
+    [Export(typeof(IApplyChangesToWorkspaceContext))]
+    internal class ApplyChangesToWorkspaceContext : OnceInitializedOnceDisposed, IApplyChangesToWorkspaceContext
+    {
+        private const string DesignTimeRuleName = CompilerCommandLineArgs.SchemaName;
+        private readonly ConfiguredProject _project;
+        private readonly ICommandLineParserService _commandLineParser;
+        private readonly IProjectLogger _logger;
+
+        private IWorkspaceProjectContext _context;
+        private ExportLifetimeContext<IWorkspaceContextHandler>[] _handlers;
+
+        [ImportingConstructor]
+        public ApplyChangesToWorkspaceContext(ConfiguredProject project, ICommandLineParserService commandLineParser, IProjectLogger logger)
+        {
+            _project = project;
+            _commandLineParser = commandLineParser;
+            _logger = logger;
+
+            WorkspaceContextHandlerFactories = new OrderPrecedenceExportFactoryCollection<IWorkspaceContextHandler>();
+        }
+
+        [ImportMany]
+        public OrderPrecedenceExportFactoryCollection<IWorkspaceContextHandler> WorkspaceContextHandlerFactories
+        {
+            get;
+        }
+
+        public void Initialize(IWorkspaceProjectContext context)
+        {
+            Requires.NotNull(context, nameof(context));
+
+            lock (SyncObject)
+            {
+                if (_context != null)
+                    throw new InvalidOperationException("Already initialized.");
+
+                _context = context;
+
+                EnsureInitialized();
+            }
+        }
+
+        public void ApplyDesignTime(IProjectVersionedValue<IProjectSubscriptionUpdate> update, bool isActiveContext, CancellationToken cancellationToken)
+        {
+            Requires.NotNull(update, nameof(update));
+            
+            lock (SyncObject)
+            {
+                VerifyInitializedAndNotDisposed();
+
+                IProjectChangeDescription projectChange = update.Value.ProjectChanges[DesignTimeRuleName];
+
+                if (projectChange.Difference.AnyChanges)
+                {
+                    IComparable version = GetConfiguredProjectVersion(update);
+
+                    ProcessOptions(projectChange.After);
```

Comment:
Consider having this code call the batching APIs that @jasonmalinowski has been working on implementing.

Reply:
Do we have a package I can consume? In what release are these checked in?

Reply:
They aren't yet - I was testing for the and calling via reflection if they are present.

@jasonmalinowski maybe we should add the API and just have it be a no-op for now?